python/GafferUI/EditMenu.py

```diff
@@ -192,6 +193,72 @@ def paste( menu ) :
 
 		s.graphEditor.frame( s.script.selection(), extend = True )
 
+def duplicateWithInputs( menu ) :
+
+	s = scope( menu )
+
+	errorHandler = GafferUI.ErrorDialogue.ErrorHandler(
+		title = "Errors Occurred During Duplication",
+		closeLabel = "Oy vey",
+		parentWindow = s.scriptWindow
+	)
+
+	with Gaffer.UndoScope( s.script ), errorHandler :
+		source = s.script.serialise( s.parent, s.script.selection() )
+
+		# Paste the selected nodes into an empty script to determine the order
+		# in which they are serialized.
+		dummyNodes = []
+		dummyScript = Gaffer.ScriptNode()
+		caDummy = dummyScript.childAddedSignal().connect( lambda parent, child : dummyNodes.append( child ), scoped = True )
+		dummyScript.execute( source )
+
+		# Paste them again into the destination script, which will create them
+		# in the same order as `dummyScript`, but with different names due to
+		# name collision handling.
+		newNodes = []
+		caScript = s.parent.childAddedSignal().connect( lambda parent, child : newNodes.append( child ), scoped = True )
+		s.script.execute( source, s.parent )
+
+		def duplicateInputs( source, dest ) :
+
+			if isinstance( dest, Gaffer.Plug ) :
+				if dest.direction == Gaffer.Plug.Direction.Out :
+					return
+
+				if source.getInput() is not None and dest.getInput() is None:
+					if isinstance( source.getInput().node(), Gaffer.Expression ) :
+						# If the source input is an expression, we'd be connecting
+						# to a potentially ephemeral plug on the expression that may
+						# change with expression output changes. We connect to the source
+						# directly for better reliability at the expense of consistency.
+						dest.setInput( source )
+					else :
+						dest.setInput( source.getInput() )
+					return
+
+			for i in range( 0, len( source.children() ) ) :
+				duplicateInputs( source[i], dest[i] )
+
+		# The name of a dummy node relative to `dummyScript` will match the
+		# name of the source node relative to the scoped parent from which
+		# to duplicate input connections.
+		for i in range( 0, len( newNodes ) ) :
+			duplicateInputs(
+				s.parent.descendant( dummyNodes[i].relativeName( dummyScript ) ),
+				newNodes[i]
+			)
+
+		s.script.selection().clear()
+		for n in newNodes :
+			s.script.selection().add( n )
+			p = s.graphEditor.graphGadget().getNodePosition( n )
+			offset = s.graphEditor.graphGadget().nodeGadget( n ).bound().size().y * 0.5
```

Comment:
We need to use exactly the same offset for all the nodes, otherwise they lose their positions relative to one another. Here's an extreme example where I duplicated a backdrop with a node in it :

![image](https://user-images.githubusercontent.com/1133871/186931420-5aae7ad1-72cc-4d14-b523-de9dd211da7e.png)

I think it would be fine to use a constant offset at all times, with whatever value `bound().size().y * 0.5` gives for a standard node.

Reply:
Fixed in 60dc668427f0fe3265e595d415d647521380ddfa